### PR TITLE
Add daemon for mirroring the artifacts

### PR DESCRIPTION
**Pull request by klausenbusk**
syncd is a daemon for mirroring artifacts described in a manifest file
created by the syncctl tool, between two sets of artifacts services
(registry, chart repository, yggdrasil git repository).

The goal is automating the last mile of the deployment.

### CODE REVIEW

File: syncd/main.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	return lastSuccessfulSynchronization, nil
+}
+
+func main() {
+	flag.Parse()
+
+	var lastSuccessfulSynchronization string
+
+	log.Printf("Synchronizing every %d second", syncInterval)
+	for range time.Tick(time.Second * time.Duration(syncInterval)) {
```
Comment (skovbakke) [inline syncd/main.go:189]: Duration counts nanoseconds

File: syncd/main.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	//mirrorImages
+	//mirrorYggdrasil
+	// TODO: return the current revision
+	return lastSuccessfulSynchronization, nil
+}
+
+func main() {
+	flag.Parse()
+
+	var lastSuccessfulSynchronization string
```
Comment (skovbakke) [inline syncd/main.go:186]: Maybe this could be retrieved from the target?

**Review by skovbakke — COMMENTED** [review]
Overall i think it looks like a good start

File: syncd/main.go (hunk 1 above)
Comment (klausenbusk) [reply syncd/main.go:193]: True, but it is multiplied with `time.Second`, so ind the end you get the duration in seconds.

File: syncd/main.go (hunk 2 above)
Comment (klausenbusk) [reply syncd/main.go:212]: Huh? We need to store it, so we don't do unnecessary synchronizations.

File: syncd/main.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+func mirrorImage(src, dst string, image image) error {
+	srcImage := fmt.Sprintf("%s/%s/%s@%s", src, image.Registry, image.Image, image.Digest)
+	var dstImage string
+	if image.Tag == "" {
+		dstImage = srcImage
+	} else {
+		dstImage = fmt.Sprintf("%s/%s/%s:%s", dst, image.Registry, image.Image, image.Tag)
+	}
+
+	cmd := exec.Command("skopeo", "copy", "--all", "--preserve-digests", "--src-tls-verify=false", "--dest-tls-verify=false", fmt.Sprintf("docker://%s", srcImage), fmt.Sprintf("docker://%s", dstImage))
```
Comment (MadsHT) [inline syncd/main.go:76]: Does the skopeo tool come in the Alpine img?

File: syncd/main.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("chart %v not found on %v", chart, src)
+	}
+
+	h := sha256.New()
+
+	if _, err := io.Copy(h, resp.Body); err != nil {
+		return err
+	}
+
```
Comment (MadsHT) [inline syncd/main.go:109]: I am confused about how this mirros a chart?

**Review by MadsHT — COMMENTED** [review]
Just two questions :)

File: syncd/main.go (hunk 4 above)
Comment (klausenbusk) [reply syncd/main.go:109]: It doesn't as the endpoint hasn't been defined :/

File: syncd/main.go (hunk 3 above)
Comment (klausenbusk) [reply syncd/main.go:76]: Good catch!

File: syncd/main.go (hunk 4 above)
Comment (MadsHT) [reply syncd/main.go:109]: That would explain it.. :P

Comment (klausenbusk) [reply syncd/main.go:109]: I added a `TODO` comment.

File: syncd/main.go (hunk 3 above)
Comment (klausenbusk) [reply syncd/main.go:76]: Fixed

File: syncd/main.go (hunk 4 above)
Comment (MadsHT) [reply syncd/main.go:109]: Does "how" we push it depend on what kind of destination we're pushing to?

**Review by MadsHT — APPROVED** [review]
Still a little confused about the chart dest, but otherwise LGTM

File: syncd/main.go (hunk 4 above)
Comment (klausenbusk) [reply syncd/main.go:109]: I don't think so. It should be the same kind of endpoint (could be S3, sftp etc.), just the target is different.

**Pull request merged**